File: templates/csharp/protocols/mavlink_cancellation.cs
namespace Mavlink;

/// <summary>Thrown when a MAVLink wait or long-running protocol operation is cancelled.</summary>
public sealed class MavlinkCancelledException : Exception
{
    public MavlinkCancelledException(string message = "Operation cancelled")
        : base(message)
    {
    }
}

/// <summary>
/// Cooperative cancellation token for <see cref="MavlinkSession"/> waits and protocol flows.
/// </summary>
public sealed class MavlinkCancellationToken : IDisposable
{
    private readonly object _lock = new();
    private readonly List<Action> _callbacks = new();
    private bool _cancelled;

    public bool IsCancelled
    {
        get
        {
            lock (_lock)
            {
                return _cancelled;
            }
        }
    }

    /// <summary>Fires once when <see cref="Cancel"/> is called.</summary>
    public event Action? OnCancel;

    public void Cancel()
    {
        List<Action>? callbacks;
        lock (_lock)
        {
            if (_cancelled)
            {
                return;
            }

            _cancelled = true;
            callbacks = _callbacks.ToList();
            _callbacks.Clear();
        }

        OnCancel?.Invoke();
        foreach (var callback in callbacks)
        {
            callback();
        }
    }

    public void ThrowIfCancelled()
    {
        if (IsCancelled)
        {
            throw new MavlinkCancelledException();
        }
    }

    internal IDisposable Register(Action callback)
    {
        lock (_lock)
        {
            if (_cancelled)
            {
                callback();
                return new NoopDisposable();
            }

            _callbacks.Add(callback);
            return new Registration(this, callback);
        }
    }

    public void Dispose()
    {
        lock (_lock)
        {
            _callbacks.Clear();
        }
    }

    private void Unregister(Action callback)
    {
        lock (_lock)
        {
            _callbacks.Remove(callback);
        }
    }

    private sealed class Registration(MavlinkCancellationToken owner, Action callback) : IDisposable
    {
        private bool _disposed;

        public void Dispose()
        {
            if (_disposed)
            {
                return;
            }

            _disposed = true;
            owner.Unregister(callback);
        }
    }

    private sealed class NoopDisposable : IDisposable
    {
        public void Dispose()
        {
        }
    }
}

File: templates/py/protocols/mavlink_cancellation.py
from __future__ import annotations

import asyncio
from collections.abc import Awaitable, Callable


class MavlinkCancelledException(Exception):
    def __init__(self, message: str = "Operation cancelled") -> None:
        super().__init__(message)
        self.message = message

    def __str__(self) -> str:
        return f"MavlinkCancelledException: {self.message}"


class MavlinkCancellationToken:
    """Cooperative cancellation token for session waits and protocol flows."""

    def __init__(self) -> None:
        self._cancelled = False
        self._event = asyncio.Event()
        self._callbacks: list[Callable[[], None]] = []

    @property
    def is_cancelled(self) -> bool:
        return self._cancelled

    def on_cancel(self, callback: Callable[[], None]) -> Callable[[], None]:
        """Register a callback fired once when [cancel] is called."""

        if self._cancelled:
            callback()
            return lambda: None

        self._callbacks.append(callback)

        def unsubscribe() -> None:
            try:
                self._callbacks.remove(callback)
            except ValueError:
                pass

        return unsubscribe

    def cancel(self) -> None:
        if self._cancelled:
            return
        self._cancelled = True
        self._event.set()
        for callback in list(self._callbacks):
            callback()

    def throw_if_cancelled(self) -> None:
        if self._cancelled:
            raise MavlinkCancelledException()

    async def wait_cancelled(self) -> None:
        await self._event.wait()

    def dispose(self) -> None:
        self._callbacks.clear()
